### PR TITLE
builtins: add fuzzystrmatch metaphone() built-in

### PR DESCRIPTION
**Pull request by charlespnh**
informs: https://github.com/cockroachdb/cockroach/issues/56820

Release note (sql change): The metaphone() builtin function was added.

### CODE REVIEW

**Review by rafiss — CHANGES_REQUESTED** [review]
Thanks for your contribution! can you please add tests for this in pkg/sql/logictest/testdata/logic_test/fuzzystrmatch ?

**Comment by charlespnh** [comment]
@rafiss Sorry for the delay... I intend to add tests but the built-in is a bit tricky to implement with all the rules. The PR was a draft since it was still a work in progress, but commits have been squashed and it's ready for review now. 
It seems like Postgres doesn't have many tests for metaphone. Please let me know if there are any other meaningful tests I should have added
 Both metaphone(...)'s behaviour and interface are intended to match with Postgres

**Comment by yuzefovich** [comment]
Friendly ping @rafiss

**Review by rafiss — CHANGES_REQUESTED** [review]
Thanks for your continued work on this @charlespnh!

I found a few more test cases from looking at other open-source implementations of metaphone. There are a few where this PR differs:
```
metaphone('Campbel', 10) -> expected "KMPBL" got "MBL"
metaphone('Cammmppppbbbeeelll', 10) -> expected "KMPBL" got "M"
metaphone('What', 10) -> expected "WT" got "HT"
metaphone('astronomical', 10) -> expected "ASTRNMKL" got "ASTRNML"
metaphone('district', 10) -> expected "TSTRKT" got "TSTR"
metaphone('hockey', 10) -> expected "HK" got "H"
metaphone('capital', 10) -> expected "KPTL" got "PTL"
metaphone('lightning', 10) -> expected "LTNNK" got "LFTK"
metaphone('light', 10) -> expected "LT" got "LFT"
```

I have added these test cases to your PR, but I will leave it to you to fix the remaining bugs, if you're still interested in working on this. Consider using other implementations as a reference (and cite them in comments if you do so). see https://github.com/go-dedup/metaphoneand and https://github.com/vividvilla/metaphone

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110950#-:-NmlW7JHC4uEbFsynxLl:bub41yh)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi, @jordanlewis, and @michae2)

**Review by rafiss — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110950)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @charlespnh, @fqazi, @jordanlewis, and @michae2)

___
*[`pkg/util/fuzzystrmatch/metaphone_test.go` line 90 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/110950#-NnedTuLD9hxCZpG2u5s:-NnedTuLD9hxCZpG2u5t:bn84i7l) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b37eae1822d9240c34c367b17febe639d862ed6/pkg/util/fuzzystrmatch/metaphone_test.go#L90)):*
> ```Go
> 		{
> 			Source:   "What",
> 			Expected: "HT",
> ```

postgres returns `WT` for this test case
___
*[`pkg/util/fuzzystrmatch/metaphone_test.go` line 142 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/110950#-NnedYtR2v3QMI4qSjKU:-NnedYtR2v3QMI4qSjKV:b-dnwqhz) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b37eae1822d9240c34c367b17febe639d862ed6/pkg/util/fuzzystrmatch/metaphone_test.go#L142)):*
> ```Go
> 		{
> 			Source:   "lightning",
> 			Expected: "LFTNNK",
> ```

postgres returns `LTNNK` for this case
___
*[`pkg/util/fuzzystrmatch/metaphone_test.go` line 146 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/110950#-NnedbQg6UwLMF7BhCjR:-NnedbQg6UwLMF7BhCjS:bqooiz6) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b37eae1822d9240c34c367b17febe639d862ed6/pkg/util/fuzzystrmatch/metaphone_test.go#L146)):*
> ```Go
> 		{
> 			Source:   "light",
> 			Expected: "LFT",
> ```

postgres returns `LT` for this case

**Review by charlespnh — COMMENTED** [review]
Thanks for helping me out! I'm using PostgreSQL 12.17 and this is what I get though:



> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110950#-:-Nneh_Zk29UcGyhMPnH4:bhzaax8)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi, @jordanlewis, and @michae2)

File: pkg/util/fuzzystrmatch/metaphone.go [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+// Reading, Mass: Addion-Wesley, 1995 pp160-169
+
+// Special encodings
+const SH = 'X'
+const TH = '0'
+
+/* Port from Postgres */
+var _codes = [26]int{1, 16, 4, 16, 9, 2, 4, 16, 9, 2, 0, 2, 2, 2, 1, 4, 0, 2, 4, 4, 1, 0, 0, 0, 8, 0}
+
+//					 a  b   c  d   e  f  g  h   i  j  k  l  m  n  o  p  q  r  s  t  u  v  w  x  y  z
```
Comment (rafiss) [inline pkg/util/fuzzystrmatch/metaphone.go:28]: nit: make sure the whitespace alignment doesn't use tabs here.

**Review by rafiss — CHANGES_REQUESTED** [review]
thanks for clarifying. i must have been checking against a different implementation; i confirmed that i see the same postgres behavior now.

just had one more nit

File: pkg/util/fuzzystrmatch/metaphone.go (hunk 1 above)
Comment (charlespnh) [reply pkg/util/fuzzystrmatch/metaphone.go]: done!

**Review by rafiss — APPROVED** [review]
thanks for your work on this!

bors r+

**Pull request merged**